playground/serve.go

```diff
@@ -16,27 +16,301 @@
+	// The name chosen includes the timestamp of the last known
+	// components list, which means that only one update task can be
+	// created for each version known.
+	// Where the timestamp is zero, a different value must be used as zero
+	// would remain in the dedupe list for 9 days. The startup time of the
+	// instance is used instead
+	n := timestamp
+	if timestamp == 0 {
+		n = instanceStartup
+	}
+	t.Name = fmt.Sprintf("amp-components-list-last-known-%d", n)
+	minBackoff, _ := time.ParseDuration("20s")
+	t.RetryOptions = &taskqueue.RetryOptions{
+		RetryLimit:   5,
+		MinBackoff:   minBackoff,
+		MaxDoublings: 3,
+	}
+	log.Infof(ctx, "Adding to taskqueue: %s", t.Name)
+	taskqueue.Add(ctx, t, "")
+}
+
+func componentsTask(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	_, err := fetchAndUpdateComponents(ctx)
+	if err != nil {
+		log.Warningf(ctx, "Marking task for retry, if retries remaining")
+		// Error code of 500 ensures task is marked for retry.
+		http.Error(w, "Server error", http.StatusInternalServerError)
+		return
+	}
+	w.Write([]byte("Updated components"))
+}
+
+func fetchAndUpdateComponents(ctx context.Context) ([]byte, *ComponentsReqError) {
+	log.Infof(ctx, "Fetching components list from GitHub")
+	components, err := fetchComponents(ctx)
+	if err != nil {
+		log.Warningf(ctx, "Failed to fetch components map from GitHub: %v, %v", err.Code, err.Message)
+		return nil, err
+	}
+	addComponentsToStores(ctx, components)
+	return components, nil
+}
+
+func fetchComponentsFromDataStore(ctx context.Context) (*AmpComponentsList, *ComponentsReqError) {
+	log.Infof(ctx, "Retrieving components from datastore")
+	var components AmpComponentsList
+	key := datastore.NewKey(ctx, "AmpComponentsList", "ComponentsListKey", 0, nil)
+	err := datastore.Get(ctx, key, &components)
+	if err != nil {
+		log.Warningf(ctx, "Error retrieving components from datastore")
+		rawComponents, compErr := fetchAndUpdateComponents(ctx)
+		if compErr != nil {
+			return nil, compErr
+		}
+		components.Components = rawComponents
+		components.Timestamp = int(time.Now().Unix())
+	}
+	return &components, nil
+}
+
+func fetchComponentsFromMemCache(ctx context.Context) (*AmpComponentsList, *ComponentsReqError) {
+	log.Infof(ctx, "Retrieving components from memcache")
+	var components AmpComponentsList
+	_, err := memcache.Gob.Get(ctx, COMPONENTS_MEMCACHE_KEY, &components)
+	if err == memcache.ErrCacheMiss {
+		log.Infof(ctx, "Components not in memcache, retrieving from datastore")
+		dsComponents, err := fetchComponentsFromDataStore(ctx)
+		if err == nil {
+			log.Infof(ctx, "Setting datastore components value to memcache")
+			memcacheItem := &memcache.Item{
+				Key:    COMPONENTS_MEMCACHE_KEY,
+				Object: dsComponents,
+			}
+			memcache.Gob.Set(ctx, memcacheItem)
+			return dsComponents, nil
+		}
+	} else if err != nil {
+		return nil, &ComponentsReqError{"Memcache error", http.StatusInternalServerError}
```

Comment:
We should log this error. However, we should also fallback to loading the components in this case. Suggestion: 

```
if err == nil {
  return &components, nil
}

if err == memcache.ErrCacheMiss {
   log.Infof(ctx, "Components not in memcache, retrieving from datastore")
} else {
   log.Errorf(ctx, "Error when retrieving components from memcache: %v, %v", err.Code, err.Message)
}

dsComponents, err := fetchComponentsFromDataStore(ctx)
...
```

Reply:
Great! Done